Feat: home page lists posts in pages

nauynix opened the pull request:
## Problem
Homepage posts are not paginated, leading to two issues. 
1. the size of response is quite big as server is returning whole list of post, and will only grow with number of posts
2. The page becomes very long and tedious to scroll through

- Closes #20

## Solution

This PR adds the ability to query for `posts` and `answerable posts` with two new request querys: `page` and `size`, which are optional. If they are not specified, the server returns the whole list. 

- `/posts` and `/posts/answerable` used to return `posts`, now returns shape `{posts, totalItems}`
- Take the opportunity to correct mistakes in documentation 
- Remove superseded and unused routes in `post.routes`. The functionality of `/top` and `/tag/:tagname` is covered in `/`
- Update `post` endpoint in `README`
- Update frontend to use new query, cache previous results to reduce requests when navigating back to previous page

## Screenshots
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/20250559/130751230-088f5056-8b28-438e-98df-3e36777ad6e4.png">

nauynix committed:
fix: spelling error

nauynix committed:
chore: reduce number of queries when revisiting page

nauynix committed:
feat: add pagination query to listPosts api

nauynix committed:
chore: update @access tag of post routes

nauynix committed:
chore(post): remove superseded api endpoint

nauynix committed:
Merge branch 'develop' into feat/pagination

nauynix committed:
chore(posts): rename returned object for clarity

nauynix committed:
fix: bug from JS adding strings instead of number

nauynix committed:
feat(posts): switch frontend to paginated query

LoneRifle commented on server/src/modules/post/post.service.ts:
```diff
@@ -41,17 +41,44 @@ export class PostService {
   }
 
+  private getPaginatedPosts = (
+    post: Post[],
```
nitpick
```suggestion
    posts: Post[],
```
and elsewhere

LoneRifle commented on server/src/modules/post/post.service.ts:
```diff
@@ -117,12 +144,16 @@ export class PostService {
       // posts without answers
-      return this.filterPostsWithoutAnswers(posts)
+      return this.getPaginatedPosts(
+        await this.filterPostsWithoutAnswers(posts),
```
nit - just as good practice and to make it clearer that we await on this before moving to `this.getPaginatedPosts()`, I would have this on its own line.

LoneRifle reviewed (changes requested):
Couple more points:
- Provide _some_ tests for at least post.service, at least for the functions you changed
- Current page size is hardcoded to 10, which I've personally found to be quite limiting. Raising that limit to 50 may provide a better balance of number of pages to click through vs resulting response size.

lgtm otherwise

nauynix committed:
chore(homepage): increase default page size to 50

nauynix committed:
refactor: rename key to reflect plurality

nauynix committed:
refactor(post service): seperate lines for clarity

nauynix committed:
chore: inject models into service to prepare for test coverage

nauynix committed:
chore(post): update params as optional

nauynix committed:
chore(post): prepare controller and routes for test

nauynix committed:
feat(post): add test coverage

nauynix committed:
chore(post): provide documentation

Add jsdoc and reorder the functions to be consistent
in order across route, controller and service. Renamed
some functions for consistency too.

nauynix committed:
chore: update test to reflect previous refactor

LoneRifle reviewed (approved):
nicely done
